common/src/main/java/tachyon/MasterClientBase.java

```diff
@@ -0,0 +1,190 @@
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  protected final boolean mUseZookeeper;
+  protected final ExecutorService mExecutorService;
+  protected final TachyonConf mTachyonConf;
+
+  protected InetSocketAddress mMasterAddress = null;
+  protected TProtocol mProtocol = null;
+  protected volatile boolean mConnected;
+  protected volatile boolean mIsClosed;
+
+  public MasterClientBase(InetSocketAddress masterAddress, ExecutorService executorService,
+                          TachyonConf tachyonConf) {
+    mTachyonConf = tachyonConf;
+    mUseZookeeper = mTachyonConf.getBoolean(Constants.USE_ZOOKEEPER);
+    if (!mUseZookeeper) {
+      mMasterAddress = masterAddress;
+    }
+    mConnected = false;
+    mIsClosed = false;
+    mExecutorService = executorService;
+  }
+
+  /**
+   * Returns the name of the service.
+   *
+   * @return A string representing the specific master
+   */
+  protected abstract String getServiceName();
+
+  /**
+   * This called after the connection is made to the master. Here, implementations should create
+   * internal state to finish the connection process.
+   */
+  protected abstract void afterConnect();
+
+  /**
+   * This called after the connection is disconnected. Here, implementations should clean up any
+   * additional state created for the connection.
+   */
+  protected abstract void afterDisconnect();
+
+  /**
+   * Connect with the master; an exception is thrown if this fails.
+   *
+   * @throws IOException
+   */
+  public synchronized void connect() throws IOException {
+    if (mConnected) {
+      return;
+    }
+    disconnect();
+
+    if (mIsClosed) {
+      throw new IOException("Client is closed, will not try to connect");
+    }
+
+    Exception lastException = null;
+    int maxConnectsTry = mTachyonConf.getInt(Constants.MASTER_RETRY_COUNT);
+    RetryPolicy retry = new ExponentialBackoffRetry(50, Constants.SECOND_MS, maxConnectsTry);
+    do {
+      mMasterAddress = getMasterAddress();
+
+      LOG.info("Tachyon client (version " + Version.VERSION + ") is trying to connect with "
+          + getServiceName() + " master @ " + mMasterAddress);
+
+      TProtocol binaryProtocol = new TBinaryProtocol(new TFramedTransport(
+          new TSocket(NetworkAddressUtils.getFqdnHost(mMasterAddress), mMasterAddress.getPort())));
+      mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
+      try {
+        mProtocol.getTransport().open();
+      } catch (TTransportException e) {
+        lastException = e;
+        LOG.error("Failed to connect (" + retry.getRetryCount() + ") to " + getServiceName()
+            + " master @ " + mMasterAddress + " : " + e.getMessage());
+        continue;
+      }
+
+      LOG.info("Client registered with " + getServiceName() + " master @ " + mMasterAddress);
+      mConnected = true;
+      afterConnect();
+      return;
+    } while (retry.attemptRetry() && !mIsClosed);
```

Comment:
Is it possible to do this with a `while` loop? It seems like we use `do-while` very rarely.